MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java

```diff
@@ -0,0 +1,340 @@
+        // Set up the executor and play/pause process to facilitate stopping and starting the video
+        playProcess = new PlayProcess();
+
+        surfaceHolder = this.getHolder();
+        surfaceHolder.addCallback(new SurfaceHolder.Callback() {
+            @Override
+            public void surfaceCreated(SurfaceHolder holder) {
+                Log.d(TAG, "Surface created");
+            }
+
+            @RequiresApi(api = 23)
+            @Override
+            public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
+                Log.d(TAG, "Surface changed");
+                if (savedSurface == null) {
+                    savedSurface = holder.getSurface();
+
+                    // scale the video to fit in the texture view
+                    viewWidth = width;
+                    viewHeight = height;
+                    adjustAspectRation(viewWidth, viewHeight);
+
+                    // create a new motion photo reader
+                    if (reader == null) {
+                        try {
+                            reader = MotionPhotoReader.open(filename, savedSurface);
+                            Log.d(TAG, "New motion photo reader created");
+                        } catch (IOException | XMPException e) {
+                            e.printStackTrace();
+                        }
+                    }
+
+                    // continue playing the video if not in paused state
+                    reader.seekTo(savedTimestampUs, MediaExtractor.SEEK_TO_PREVIOUS_SYNC);
+                    if (!isPaused) {
+                        executor.submit(playProcess);
+                    }
+                }
+            }
+
+            @Override
+            public void surfaceDestroyed(SurfaceHolder holder) {
+                Log.d(TAG, "Surface destroyed");
+                playProcess.cancel();
+            }
+        });
+    }
+
+    private void adjustAspectRation(int viewWidth, int viewHeight) {
+        double aspectRatio = (videoRotation > 0) ?
+                (double) videoWidth / videoHeight : (double) videoHeight / videoWidth;
+        Log.d(TAG, "asepct ratio: " + aspectRatio);
+        int newVideoWidth, newVideoHeight;
+        if (viewHeight > (int) (viewWidth * aspectRatio)) {
+            // limited by narrow width
+            if (fill) {
+                newVideoWidth = (int) (viewHeight / aspectRatio);
+                newVideoHeight = viewHeight;
+            } else {
+                newVideoWidth = viewWidth;
+                newVideoHeight = (int) (viewWidth * aspectRatio);
+            }
+        } else {
+            // limited by short height
+            if (fill) {
+                newVideoWidth = viewWidth;
+                newVideoHeight = (int) (viewWidth * aspectRatio);
+            } else{
+                newVideoWidth = (int) (viewHeight / aspectRatio);
+                newVideoHeight = viewHeight;
+            }
+        }
```

Comment:
This will be easier to read if replaced with a private static helper method.

Reply:
(Deleted this method + usages, will use OpenGL to adjust later).